jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java

```diff
@@ -42,57 +43,65 @@ public void validateFields( Object object ) {
             if( resolution == Resolution.NAME ) {
-                key = field.getName();
+                key = field.getField().getName();
             } else {
-                key = field.getType();
+                key = field.getField().getType();
             }
             if( resolvedFields.containsKey( key ) ) {
                 Field existingField = resolvedFields.get( key );
                 throw new AmbiguousResolutionException( "Ambiguous fields with same " + resolution + " detected. Field 1: " +
-                        existingField + ", field 2: " + field );
+                        existingField + ", field 2: " + field.getField() );
             }
-            resolvedFields.put( key, field );
+            resolvedFields.put( key, field.getField() );
         }
 
         validatedClasses.put( object.getClass(), Boolean.TRUE );
     }
 
-    private List<Field> getScenarioFields( Object object ) {
-        return FieldCache.get( object.getClass() ).getFieldsWithAnnotation( ScenarioState.class, ProvidedScenarioState.class,
-            ExpectedScenarioState.class );
+    private List<ScenarioStateField> getScenarioFields( Object object ) {
+        @SuppressWarnings( "unchecked" )
+        List<Field> scenarioFields = FieldCache
+            .get( object.getClass() )
+            .getFieldsWithAnnotation( ScenarioState.class, ProvidedScenarioState.class, ExpectedScenarioState.class );
+
+        return Lists.transform( scenarioFields, ScenarioStateField.fromField );
     }
 
     @SuppressWarnings( "unchecked" )
     public void readValues( Object object ) {
         validateFields( object );
-        for( Field field : getScenarioFields( object ) ) {
+        for( ScenarioStateField field : getScenarioFields( object ) ) {
             try {
-                Object value = field.get( object );
+                Object value = field.getField().get( object );
                 updateValue( field, value );
-                log.debug( "Reading value {} from field {}", value, field );
+                log.debug( "Reading value {} from field {}", value, field.getField() );
             } catch( IllegalAccessException e ) {
-                throw new RuntimeException( "Error while reading field " + field, e );
+                throw new RuntimeException( "Error while reading field " + field.getField(), e );
             }
         }
     }
 
     @SuppressWarnings( "unchecked" )
     public void updateValues( Object object ) {
         validateFields( object );
-        for( Field field : getScenarioFields( object ) ) {
-            try {
-                Object value = getValue( field );
-                if( value != null ) {
-                    field.set( object, value );
-                    log.debug( "Setting field {} to value {}", field, value );
+        for( ScenarioStateField field : getScenarioFields( object ) ) {
+            Object value = getValue( field );
+
+            if( value != null ) {
+                try {
```

Comment:
Moved these try/catch blocks around so the IllegalAccess block only encapsulates the code that actually can throw this.